File: backend/routers/party_bank_accounts.py
"""Iter150G · Party Bank Account router (new · non-locked)."""
from __future__ import annotations
from fastapi import APIRouter, Depends, HTTPException, Query, Request

from db import db
from models_bank import PartyBankAccount, PARTY_TYPES
from auth import get_current_user
from company import _active_company_id
from audit import _log_audit
from services_bank_accounts import (
    mask_account_number, duplicate_exists_party, clear_primary_party,
    strip_full_number, party_role_can_manage, party_role_can_view_full,
)
from models import now_utc, new_id

router = APIRouter(prefix="/api", tags=["PartyBankAccount"])

_PARTY_COLL = {
    "supplier": "suppliers", "vendor": "vendors", "mechanic": "mechanics",
    "driver": "drivers", "customer": "customers",
}


def _require_manage(user):
    if not party_role_can_manage(user):
        raise HTTPException(status_code=403,
                            detail="Owner or Accountant role required.")


async def _ensure_party_exists(uid: str, cid: str, party_type: str, party_id: str):
    if party_type not in PARTY_TYPES:
        raise HTTPException(400, f"unsupported party_type. Allowed: {PARTY_TYPES}")
    coll = _PARTY_COLL[party_type]
    q = {"id": party_id, "user_id": uid}
    # Company-scope only where the party master has company_id (drivers may not)
    if party_type in ("supplier", "vendor", "mechanic", "customer"):
        q["company_id"] = cid
    doc = await db[coll].find_one(q, {"_id": 0, "id": 1})
    if not doc:
        raise HTTPException(404, f"{party_type} not found in this tenant")


@router.get("/party-bank-accounts")
async def list_party_bank_accounts(
    request: Request,
    party_type: str = Query(...),
    party_id: str = Query(...),
    user=Depends(get_current_user),
):
    uid = user["user_id"]
    cid = await _active_company_id(request, user)
    if party_type not in PARTY_TYPES:
        raise HTTPException(400, f"unsupported party_type. Allowed: {PARTY_TYPES}")
    allow_full = party_role_can_view_full(user)
    rows = await db.party_bank_accounts.find(
        {"user_id": uid, "party_type": party_type, "party_id": party_id},
        {"_id": 0, "user_id": 0},
    ).sort("created_at", -1).to_list(500)
    return [strip_full_number(r, allow_full) for r in rows]


@router.post("/party-bank-accounts")
async def create_party_bank_account(
    payload: PartyBankAccount, request: Request,
    user=Depends(get_current_user),
):
    _require_manage(user)
    uid = user["user_id"]
    cid = await _active_company_id(request, user)
    await _ensure_party_exists(uid, cid, payload.party_type, payload.party_id)
    if not payload.account_number.strip() or not payload.ifsc.strip():
        raise HTTPException(400, "account_number and ifsc are required")
    if await duplicate_exists_party(uid, payload.party_type, payload.party_id,
                                     payload.account_number, payload.ifsc):
        raise HTTPException(409, "Active bank account with same (account_number, ifsc) already exists for this party")
    doc = payload.model_dump()
    doc["id"] = doc.get("id") or new_id("pba_")
    doc["user_id"] = uid
    doc["created_by"] = uid
    doc["created_at"] = now_utc().isoformat()
    doc["masked_display"] = mask_account_number(doc["account_number"])
    if doc.get("is_primary"):
        await clear_primary_party(uid, doc["party_type"], doc["party_id"],
                                  keep_id=doc["id"])
    await db.party_bank_accounts.insert_one(doc)
    doc.pop("_id", None)
    await _log_audit(
        {"user_id": uid, "company_id": cid,
         "email": user.get("email", ""), "name": user.get("name", "")},
        "party_bank_account", "create", doc["id"],
        f"{doc['party_type']}:{doc['party_id']}", "", {},
    )
    return strip_full_number(doc, party_role_can_view_full(user))


@router.put("/party-bank-accounts/{bid}")
async def update_party_bank_account(
    bid: str, payload: dict, request: Request,
    user=Depends(get_current_user),
):
    _require_manage(user)
    uid = user["user_id"]
    cid = await _active_company_id(request, user)
    before = await db.party_bank_accounts.find_one(
        {"id": bid, "user_id": uid}, {"_id": 0})
    if not before:
        raise HTTPException(404, "Bank account not found")
    ALLOWED = {"account_holder_name", "bank_name", "branch", "account_type",
               "verification_status"}
    patch = {k: v for k, v in (payload or {}).items() if k in ALLOWED}
    if not patch:
        raise HTTPException(400, "No editable fields provided")
    patch["modified_by"] = uid
    patch["modified_at"] = now_utc().isoformat()
    await db.party_bank_accounts.update_one({"id": bid}, {"$set": patch})
    await _log_audit(
        {"user_id": uid, "company_id": cid,
         "email": user.get("email", ""), "name": user.get("name", "")},
        "party_bank_account", "edit", bid,
        f"{before['party_type']}:{before['party_id']}", "",
        {k: {"old": before.get(k), "new": patch.get(k)} for k in patch
         if k not in ("modified_by", "modified_at")},
    )
    return strip_full_number({**before, **patch}, party_role_can_view_full(user))


@router.post("/party-bank-accounts/{bid}/set-primary")
async def set_primary_party_ep(
    bid: str, request: Request, user=Depends(get_current_user),
):
    _require_manage(user)
    uid = user["user_id"]
    doc = await db.party_bank_accounts.find_one(
        {"id": bid, "user_id": uid, "is_active": True}, {"_id": 0})
    if not doc:
        raise HTTPException(404, "Active bank account not found")
    await clear_primary_party(uid, doc["party_type"], doc["party_id"], keep_id=bid)
    await db.party_bank_accounts.update_one(
        {"id": bid}, {"$set": {"is_primary": True,
                                "modified_by": uid,
                                "modified_at": now_utc().isoformat()}},
    )
    return {"ok": True, "id": bid, "is_primary": True}


@router.post("/party-bank-accounts/{bid}/deactivate")
async def deactivate_party(
    bid: str, request: Request, payload: dict = None,
    user=Depends(get_current_user),
):
    _require_manage(user)
    uid = user["user_id"]
    cid = await _active_company_id(request, user)
    doc = await db.party_bank_accounts.find_one(
        {"id": bid, "user_id": uid}, {"_id": 0})
    if not doc:
        raise HTTPException(404, "Bank account not found")
    reason = ((payload or {}).get("reason") or "").strip()[:240]
    await db.party_bank_accounts.update_one(
        {"id": bid},
        {"$set": {"is_active": False, "is_primary": False,
                   "deactivated_by": uid,
                   "deactivated_at": now_utc().isoformat(),
                   "deactivation_reason": reason,
                   "modified_by": uid,
                   "modified_at": now_utc().isoformat()}},
    )
    await _log_audit(
        {"user_id": uid, "company_id": cid,
         "email": user.get("email", ""), "name": user.get("name", "")},
        "party_bank_account", "deactivate", bid,
        f"{doc['party_type']}:{doc['party_id']}", reason, {},
    )
    return {"ok": True, "id": bid, "is_active": False}


@router.post("/party-bank-accounts/{bid}/replace")
async def replace_party(
    bid: str, payload: PartyBankAccount, request: Request,
    user=Depends(get_current_user),
):
    _require_manage(user)
    uid = user["user_id"]
    cid = await _active_company_id(request, user)
    old = await db.party_bank_accounts.find_one(
        {"id": bid, "user_id": uid, "is_active": True}, {"_id": 0})
    if not old:
        raise HTTPException(404, "Active bank account not found")
    if payload.party_type != old["party_type"] or payload.party_id != old["party_id"]:
        raise HTTPException(400, "Replacement must retain same party_type + party_id")
    new_doc = payload.model_dump()
    new_doc["id"] = new_id("pba_")
    new_doc["user_id"] = uid
    new_doc["created_by"] = uid
    new_doc["created_at"] = now_utc().isoformat()
    new_doc["is_active"] = True
    new_doc["masked_display"] = mask_account_number(new_doc["account_number"])
    if new_doc.get("is_primary") or old.get("is_primary"):
        await clear_primary_party(uid, new_doc["party_type"],
                                  new_doc["party_id"], keep_id=new_doc["id"])
        new_doc["is_primary"] = True
    await db.party_bank_accounts.insert_one(new_doc)
    new_doc.pop("_id", None)
    await db.party_bank_accounts.update_one(
        {"id": bid},
        {"$set": {"is_active": False, "is_primary": False,
                   "deactivated_by": uid,
                   "deactivated_at": now_utc().isoformat(),
                   "deactivation_reason": f"replaced_by:{new_doc['id']}"}},
    )
    await _log_audit(
        {"user_id": uid, "company_id": cid,
         "email": user.get("email", ""), "name": user.get("name", "")},
        "party_bank_account", "replace", bid,
        f"{old['party_type']}:{old['party_id']}",
        f"replaced_by:{new_doc['id']}", {},
    )
    return strip_full_number(new_doc, party_role_can_view_full(user))


@router.get("/party-bank-accounts/{bid}/reveal")
async def reveal_party_full(
    bid: str, request: Request, user=Depends(get_current_user),
):
    if not party_role_can_view_full(user):
        raise HTTPException(403, "Not permitted to reveal full account number")
    uid = user["user_id"]
    cid = await _active_company_id(request, user)
    doc = await db.party_bank_accounts.find_one(
        {"id": bid, "user_id": uid}, {"_id": 0})
    if not doc:
        raise HTTPException(404, "Bank account not found")
    await _log_audit(
        {"user_id": uid, "company_id": cid,
         "email": user.get("email", ""), "name": user.get("name", "")},
        "party_bank_account", "reveal", bid,
        f"{doc['party_type']}:{doc['party_id']}", "", {},
    )
    return {"id": bid, "account_number": doc.get("account_number", "")}

File: backend-node/src/routes/party-bank-accounts.ts
import type { FastifyInstance, FastifyReply, FastifyRequest } from 'fastify';
import type { Db, Filter, FindCursor } from 'mongodb';
import { authenticate, type AuthUser } from '../auth.js';
import { activeCompanyId } from '../tenant.js';
import { HttpError } from '../errors.js';

// ── Gate 9e · Motor `to_list(n)` (route-local) ────────────────────────
// Motor sends NO server-side limit for `to_list(n)`; a `.limit(n)` turns the
// sort into a top-k whose tie order at the cap differs (verified live). Read
// the sorted cursor and stop after n documents, exactly like Motor.
async function motorToList<T>(cursor: FindCursor<T>, n: number): Promise<T[]> {
  const out: T[] = [];
  try {
    if (n <= 0) {
      await cursor.hasNext(); // to_list(0) still issues the find, returns []
      return out;
    }
    for await (const doc of cursor) {
      out.push(doc);
      if (out.length >= n) break;
    }
  } finally {
    await cursor.close();
  }
  return out;
}

/**
 * TRUKVIA · Phase-3 · Gate-6r · Party bank account read-only shadow.
 *
 *   GET /api/party-bank-accounts?party_type=&party_id=
 *
 * Faithful shadow of:
 *   backend/routers/party_bank_accounts.py::list_party_bank_accounts
 *   (lines 43–59)
 *
 * STRICTLY READ-ONLY. Class-C — the Python handler executes ONLY
 * `db.party_bank_accounts.find().sort().to_list()` plus a per-row
 * `strip_full_number(...)` transform. Zero writer hook, zero audit
 * call, zero backfill, zero recompute, zero FinTxn emission, zero
 * approvals / policy_changes / counters / idempotency touch, zero
 * unrelated collection reads.
 *
 * ── GATE-6r NEW DIMENSIONS (BIND PRECISELY) ─────────────────────────
 *
 *   1. REQUIRED QUERY PARAMS — both `party_type` and `party_id` are
 *      required. Missing → HTTP 422 with FastAPI/Pydantic-v2 detail
 *      shape reproduced verbatim below.
 *
 *   2. TUPLE-MEMBERSHIP 400 — invalid `party_type` → HTTP 400 with
 *      EXACT literal (single-quoted tuple repr, mandatory space after
 *      each comma):
 *        "unsupported party_type. Allowed: ('supplier', 'vendor',
 *         'mechanic', 'driver', 'customer')"
 *
 *   3. AUTH-ORDER SHORT-CIRCUIT — authentication runs BEFORE query
 *      validation. Unauthenticated / invalid / expired requests
 *      surface the locked 401 literal without ever computing the
 *      422/400 shape. Confirmed against the running Python backend
 *      during Gate 6r Pre-flight.
 *
 *   4. NO `company_id` IN FILTER — the party master carries the
 *      tenant discriminator (via `party_type + party_id`), so the
 *      Python route deliberately omits `company_id` from the DB
 *      filter. `activeCompanyId()` is STILL invoked to preserve the
 *      exact tenant-resolver DB touch surface (Python binds `cid`
 *      but never consumes it).
 *
 *   5. GATE-6m MASKING SEMANTICS REUSED — the three inline helpers
 *      (`partyRoleCanViewFull`, `maskAccountNumber`, `stripFullNumber`)
 *      are duplicated verbatim from
 *      `services_bank_accounts.py::party_role_can_view_full /
 *       mask_account_number / strip_full_number` per the six-file
 *      ceiling. Locked Gate 6m file NOT modified.
 *
 * Gate-7 writer boundary (out of scope for Gate 6r):
 *   POST / PUT / set-primary / deactivate / replace / reveal
 *   (which writes an audit log) remain Python-authoritative.
 */

// ── PARTY_TYPES tuple parity (bind exact Python repr) ─────────────────
// Python: `PARTY_TYPES = ("supplier", "vendor", "mechanic", "driver", "customer")`
// f-string: `f"unsupported party_type. Allowed: {PARTY_TYPES}"` →
//   `unsupported party_type. Allowed: ('supplier', 'vendor', 'mechanic', 'driver', 'customer')`
// This string is BYTE-critical for parity — never mutate its formatting.
const PARTY_TYPES: readonly string[] = [
  'supplier', 'vendor', 'mechanic', 'driver', 'customer',
];
const PARTY_TYPES_REPR = "('supplier', 'vendor', 'mechanic', 'driver', 'customer')";

// ── Privileged-role predicate (verbatim Python semantics inline) ──────
// Python: role = (user.get("effective_role") or user.get("role") or "").lower()
// AuthUser already coalesces effective_role from session.effective_role ||
// session.role at Gate-2 auth.ts:87. The trailing `|| ''` mirrors Python's
// falsy-fallback to empty string when both are missing.
const PRIVILEGED_ROLES = new Set(['owner', 'accountant', 'admin']);

function partyRoleCanViewFull(user: AuthUser): boolean {
  const raw = user.effective_role || '';
  return PRIVILEGED_ROLES.has(raw.toLowerCase());
}

// ── mask_account_number (verbatim Python inline) ──────────────────────
function maskAccountNumber(acc: unknown): string {
  const s = (typeof acc === 'string' ? acc : '').trim();
  if (!s) return '';
  if (s.length <= 4) return 'X'.repeat(s.length);
  return 'X'.repeat(s.length - 4) + s.slice(-4);
}

// ── strip_full_number (verbatim Python inline) ────────────────────────
function stripFullNumber(
  doc: Record<string, unknown>,
  allowFull: boolean,
): Record<string, unknown> {
  if (!doc) return doc;
  const out: Record<string, unknown> = { ...doc };
  if (!allowFull) {
    const md = out['masked_display'];
    out['account_number'] = md ? md : maskAccountNumber(out['account_number']);
  }
  return out;
}

// ── Query helpers ─────────────────────────────────────────────────────
function firstStr(v: unknown): string | undefined {
  if (typeof v === 'string') return v;
  if (Array.isArray(v)) {
    const first = v[0];
    return typeof first === 'string' ? first : undefined;
  }
  return undefined;
}

// FastAPI/Pydantic-v2 "missing required query param" detail entry.
// Observed live against the running Python backend (Pre-flight parity
// harness — see /tmp/gate6r_parity_results.json.observed_py_422):
//   Python (Pydantic 2.13) emits:
//     { type: "missing", loc: ["query", <field>], msg: "Field required",
//       input: null, url: "https://errors.pydantic.dev/2.13/v/missing" }
// All five keys are reproduced verbatim below. The `url` value is
// Pydantic-minor-version-dependent; upgrading Pydantic requires a
// coordinated one-line refresh here.
function pydanticMissingEntry(field: 'party_type' | 'party_id'): Record<string, unknown> {
  return {
    type: 'missing',
    loc: ['query', field],
    msg: 'Field required',
    input: null,
    url: 'https://errors.pydantic.dev/2.13/v/missing',
  };
}

interface PartyBankAccountDoc {
  id?: string;
  user_id?: string;
  party_type?: string;
  party_id?: string;
  account_number?: string;
  masked_display?: string;
  created_at?: string;
  [k: string]: unknown;
}

// ── Handler registrar ─────────────────────────────────────────────────
export async function registerPartyBankAccountsReadRoutes(
  app: FastifyInstance,
  deps: { db: Db },
): Promise<void> {
  const { db } = deps;

  // migration-allowlisted: phase-3-gate-6r (read-only)
  app.get(
    '/api/party-bank-accounts',
    async (req: FastifyRequest, reply: FastifyReply) => {
      // 1. Auth FIRST — 401 short-circuits 422/400 (auth-order parity).
      let user: AuthUser;
      try {
        user = await authenticate(req, db);
      } catch (err) {
        if (err instanceof HttpError) {
          reply.code(err.status);
          return { detail: err.detail };
        }
        throw err;
      }
      const uid = user.user_id;

      // 2. Active-company invocation-parity (cid is intentionally UNUSED
      //    downstream — mirrors Python which binds but does not consume).
      await activeCompanyId(req, uid, db);

      // 3. Required-param validation → FastAPI/Pydantic-v2 422.
      const q = (req.query ?? {}) as Record<string, unknown>;
      const partyType = firstStr(q['party_type']);
      const partyId = firstStr(q['party_id']);
      const missing: Record<string, unknown>[] = [];
      if (partyType === undefined) missing.push(pydanticMissingEntry('party_type'));
      if (partyId === undefined) missing.push(pydanticMissingEntry('party_id'));
      if (missing.length > 0) {
        reply.code(422);
        return { detail: missing };
      }

      // 4. Tuple-membership 400 (byte-exact Python literal).
      if (!PARTY_TYPES.includes(partyType!)) {
        reply.code(400);
        return { detail: `unsupported party_type. Allowed: ${PARTY_TYPES_REPR}` };
      }

      // 5. Compute masking predicate ONCE per request.
      const allowFull = partyRoleCanViewFull(user);

      // 6. Read with EXACT filter/projection/sort/cap. NO company_id.
      const rows = await motorToList(
        db
        .collection<PartyBankAccountDoc>('party_bank_accounts')
        .find(
          {
            user_id: uid,
            party_type: partyType!,
            party_id: partyId!,
          } as Filter<PartyBankAccountDoc>,
          { projection: { _id: 0, user_id: 0 } },
        )
        .sort([['created_at', -1]]),
        500,
      );

      // 7. Apply per-row masking. Never mutates the source row.
      return rows.map((r) => stripFullNumber(r, allowFull));
    },
  );
}
